### PR TITLE
Walk/pretty print comments

### PR DESCRIPTION
**Pull request by xeger**
**NB:** This is a port of [cucumber/common#1738](https://github.com/cucumber/common/pull/1738) authored by @aslakhellesoy. I've patched the changes over from the monorepo to encourage conversation, but I am not the author nor am I intimately familiar with the problem.

## Summary
* Walk comments in `walkGherkinDocument`
* Pretty print comments in `pretty`

## Details
Keep a stack of comments and pop/walk comments until the current AST node.

## How Has This Been Tested?
I've just added a couple of unit tests - we need more thorough tests for this:

* [ ]  Add comments everywhere in `walkGherkinDocumentTest`
* [ ]  Add more edge case tests for `pretty`

## Types of changes
* [x]  Bug fix (non-breaking change which fixes an issue).
* [ ]  New feature (non-breaking change which adds functionality).
* [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:
* [ ]  The change has been ported to Java.
* [ ]  The change has been ported to Ruby.
* [x]  The change has been ported to JavaScript.
* [ ]  The change has been ported to Go.
* [ ]  The change has been ported to .NET.
* [ ]  I've added tests for my code.
* [ ]  My change requires a change to the documentation.
* [ ]  I have updated the documentation accordingly.
* [ ]  I have updated the CHANGELOG accordingly.

### CODE REVIEW

**Comment by xeger** [comment]
@aslakhellesoy I followed your advice and fixed some of the bugs people were complaining about. There are some odd cases. For example, a comment above a `"""` is preserved whereas comments above a table (but below its parent) are stripped - and I'm not sure if there _is_ any way to preserve them (nor do I care).

Another common case is block-commenting an entire section of the file while running tests interactively. I would probably add `pretty` cases to verify that works well.

**Comment by mpkorstanje** [comment]
Clicked the wrong button. Approved CI to run and turned PR back into Draft.

**Comment by mpkorstanje** [comment]
We are a bit short handed lately, I'm not sure when someone will be able to review this. I have flagged it for extra attention though.

**Comment by xeger** [comment]
I realized that the `pretty` round-tripping behavior that I thought was buggy, was actually correct. This PR is ready to go IMHO. Happy to add more test cases if anyone has recommendations.

**Comment by xeger** [comment]
Sorry to pester @mpkorstanje, but it's been a few months and my organization, plus a few others, have been happily using these `pretty` improvements via a [forked build[(https://github.com/xeger/cucumber-vscode/releases/tag/v1.7.0) of the Cucumber VS Code plugin. I'd say this counts as reasonable QA on the stuff that Aslak started and I tweaked. 😄

I'm sure there are corner cases to be handled, and nobody in my sphere is blocking (we're all using the forked build) - but in the interest of enhancing the cucumber org's visibility, y'all might want to consider merging this stuff and kicking out new releases of the plugin and its dependent libraries.

**Comment by mpkorstanje** [comment]
Currently everyone working on the project is a volunteer. It would be worth reading: https://mattwynne.net/new-beginning for a little bit more context. Unfortunately the public response from Smart Bear so far has been a deafening silence.

From a more practical point of view. We currently don't have anyone actively maintaining the VSCode plugin in the Cucumber org. Since you do have that motivation, would you be interested in turning the fork into the "official" version?

It would probably be easiest if you pop into Slack and/or join the weekly conference call to organize that. We are running a pretty informal process.

**Comment by xeger** [comment]
Thank you for the context. I'm leaning toward yes. I'll do some reading over the weekend and if I can commit, I will come say hi. 🙇

File: javascript/test/prettyTest.ts [hunk 1]
```diff
@@ -183,6 +209,7 @@ Feature: hello
 `)
   })
 
+  // TODO: locate test data outside of monorepo?
```
Comment (mpkorstanje) [inline javascript/test/prettyTest.ts:212]: Would you happen to know how this passes CI?

**Review by mpkorstanje — APPROVED** [review]
I don't understand the PR but the tests look good so I'm going to approve this on that basis.

I'll leave the PR open so I've got an ear mark to release it. Probably next Thursday pending availability.

**Comment by mpkorstanje** [comment]
> Thank you for the context. I'm leaning toward yes. I'll do some reading over the weekend and if I can commit, I will come say hi. 🙇

Cheers. Appreciated. 

Commitment wise it would be comparable to running the current fork. The release process is automated. Everything else should be more or less comparable.

**Comment by mattwynne** [comment]
> > Thank you for the context. I'm leaning toward yes. I'll do some reading over the weekend and if I can commit, I will come say hi. 🙇
> 
> Cheers. Appreciated.
> 
> Commitment wise it would be comparable to running the current fork. The release process is automated. Everything else should be more or less comparable.

👍 let us know if you have any questions @xeger!

**Comment by xeger** [comment]
I've copied `gherkin/testdata/**/*.feature` from an old commit of the monorepo and fixed the relative paths so that `prettyTest` iterates over all of the `good` examples. Happily, everything is still green.

This PR is ready to go.

**Comment by aslakhellesoy** [comment]
Hi @xeger,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber
